app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java

```diff
@@ -250,80 +251,113 @@ private void onUpdatedShouldRunDecision(boolean newShouldRunDecision) {
-                mApi = new RestApi(SyncthingService.this, mConfig.getWebGuiUrl(), mConfig.getApiKey(),
-                                    SyncthingService.this::onApiAvailable, () -> onServiceStateChange(mCurrentState));
-                Log.i(TAG, "Web GUI will be available at " + mConfig.getWebGuiUrl());
-            }
-
-            // Start the syncthing binary.
-            if (mSyncthingRunnable != null || mSyncthingRunnableThread != null) {
-                Log.e(TAG, "StartupTask/onPostExecute: Syncthing binary lifecycle violated");
+    private void launchStartupTask () {
+        Log.v(TAG, "Starting syncthing");
+        synchronized(mStateLock) {
+            if (mCurrentState != State.INIT) {
+                Log.e(TAG, "launchStartupTask: Wrong state " + mCurrentState + " detected. Cancelling.");
                 return;
             }
-            mSyncthingRunnable = new SyncthingRunnable(SyncthingService.this, SyncthingRunnable.Command.main);
-            mSyncthingRunnableThread = new Thread(mSyncthingRunnable);
-            mSyncthingRunnableThread.start();
+        }
 
-            /**
-             * Wait for the web-gui of the native syncthing binary to come online.
-             *
-             * In case the binary is to be stopped, also be aware that another thread could request
-             * to stop the binary in the time while waiting for the GUI to become active. See the comment
-             * for SyncthingService.onDestroy for details.
-             */
-            if (mPollWebGuiAvailableTask == null) {
-                mPollWebGuiAvailableTask = new PollWebGuiAvailableTask(
-                        SyncthingService.this,
-                        getWebGuiUrl(),
-                        mConfig.getApiKey(),
-                        result -> {
-                    Log.i(TAG, "Web GUI has come online at " + mConfig.getWebGuiUrl());
-                    if (mApi != null) {
-                        mApi.readConfigFromRestApi();
-                    }
-                });
-            }
+        // Safety check: Log warning if a previously launched startup task did not finish properly.
+        if (mStartupTask != null && (mStartupTask.getStatus() == AsyncTask.Status.RUNNING)) {
+            Log.w(TAG, "launchStartupTask: StartupTask is still running. Skipped starting it twice.");
+            return;
         }
+        onServiceStateChange(State.STARTING);
+        mStartupTask = new StartupTask(this);
+        mStartupTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
     }
 
     /**
-     * Called when {@link PollWebGuiAvailableTask} confirmed the REST API is available.
-     * We can assume mApi being available under normal conditions.
+     * Sets up the initial configuration, and updates the config when coming from an old
+     * version.
+     */
+     private static class StartupTask extends AsyncTask<Void, Void, Void> {
+
+         private WeakReference<SyncthingService> refSyncthingService;
```

Comment:
Noop blank line after the method signature?
I am pretty sure if you press Ctrl + Alt + L in Android studio, nitpicks like this get fixed automatically.

Reply:
removed blank line. Thanks for the hint.